sdk/clientcore/tools/annotation-processor/src/main/java/io/clientcore/tools/codegen/utils/PathBuilder.java

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.tools.codegen.utils;
+
+import io.clientcore.tools.codegen.exceptions.MissingSubstitutionException;
+import io.clientcore.tools.codegen.models.HttpRequestContext;
+import io.clientcore.tools.codegen.models.Substitution;
+
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class PathBuilder {
+    // this class takes a 'raw host' string that contains {} delimited parameters, and needs to convert it into a
+    // Java string concatenation that can be used in the generated code. For example, the raw host string:
+    // https://{endpoint}/keys/{path1}
+    // would be converted into:
+    // "https://" + endpointParam + "/keys/" + pathValue
+    // Note that query parameters may also exist, and should be appended to the end of the URL string using
+    // a Map containing key-value pairs.
+    // Note that the 'endpoint' parameter is special - it is always the first parameter, and is always a host parameter.
+    public static String buildPath(String rawHost, HttpRequestContext method) {
+        if (method == null) {
+            throw new NullPointerException("method cannot be null");
+        }
+
+        final boolean hasQueryParams = !method.getQueryParams().isEmpty();
+
+        Pattern pattern = Pattern.compile("\\{(.+?)\\}");
+        Matcher matcher = pattern.matcher(rawHost);
+        StringBuffer buffer = new StringBuffer();
+
+        while (matcher.find()) {
+            String paramName = matcher.group(1);
+            Substitution substitution = method.getSubstitution(paramName);
```

Comment:
Real niche corner case if I'm understanding this code correctly. If this is handling host and path substitutions at the same time it's possible that the host and path have substitutions with the same name:

```
@ServiceInterface(name = "HostSubstitutionMethods", host = "https://{sub1}.host.com")
interface WeirdDesign {
  @HttpRequestInformation(method = HttpMethod.GET, path = "{sub1}")
  void substitution(@HostParam("sub1") String hostSub1, @PathParam("sub1") String pathSub1);
}
```

Reply:
Let me add this to the test case and validate.